Source/JavaScriptCore/runtime/ShadowRealmPrototype.cpp

```diff
@@ -0,0 +1,127 @@
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "ShadowRealmPrototype.h"
+
+#include "Completion.h"
+#include "JSCBuiltins.h"
+#include "JSCJSValue.h"
+#include "JSGlobalObject.h"
+#include "JSModuleLoader.h"
+#include "JSInternalPromise.h"
+#include "ShadowRealmObject.h"
+
+#include "ShadowRealmPrototype.lut.h"
+
+namespace JSC {
+
+/* Source for ShadowRealmPrototype.lut.h
+@begin shadowRealmPrototypeTable
+  evaluate    JSBuiltin     DontEnum|Function  1
+  importValue JSBuiltin     DontEnum|Function  2
+@end
+*/
+
+const ClassInfo ShadowRealmPrototype::s_info = { "ShadowRealm", &Base::s_info, &shadowRealmPrototypeTable, nullptr, CREATE_METHOD_TABLE(ShadowRealmPrototype) };
+
+ShadowRealmPrototype::ShadowRealmPrototype(VM& vm, Structure* structure)
+    : JSNonFinalObject(vm, structure)
+{
+}
+
+void ShadowRealmPrototype::finishCreation(VM& vm)
+{
+    Base::finishCreation(vm);
+    ASSERT(inherits(vm, info()));
+    JSC_TO_STRING_TAG_WITHOUT_TRANSITION();
+}
+
+JSC_DEFINE_HOST_FUNCTION(importInRealm, (JSGlobalObject* globalObject, CallFrame* callFrame))
+{
+    VM& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    if (callFrame->argumentCount() != 2) {
+        throwTypeError(globalObject, scope, "Expected to be called with two arguments"_s);
+        RELEASE_AND_RETURN(scope, JSValue::encode(jsUndefined()));
+    }
+
+    JSValue thisValue = callFrame->uncheckedArgument(0);
+    ShadowRealmObject* thisRealm = jsDynamicCast<ShadowRealmObject*>(vm, thisValue);
+    if (UNLIKELY(!thisRealm))
+        return throwVMTypeError(globalObject, scope, "'this' should be a ShadowRealm");
+
+    auto* promise = JSPromise::create(vm, globalObject->promiseStructure());
+
+    auto sourceOrigin = callFrame->callerSourceOrigin(vm);
+    auto* specifier = callFrame->uncheckedArgument(1).toString(globalObject);
+    RETURN_IF_EXCEPTION(scope, { });
+
+    JSGlobalObject* realmGlobalObject = thisRealm->globalObject();
+    auto* internalPromise = realmGlobalObject->moduleLoader()->importModule(realmGlobalObject, specifier, jsUndefined(), sourceOrigin);
```

Comment:
this creates an internal promise somehow associated with the realm global object, but we later do `resolve` on the original global object. Felt worth double-checking that the difference is fine